client/api/src/notifications/registry.rs

```diff
@@ -0,0 +1,379 @@
+	fn drop(&mut self) {
+		if !self.was_triggered {
+			log::trace!(
+				target: "storage_notifications",
+				"Listener was never triggered: id={}, keys={:?}, child_keys={:?}",
+				self.subs_id,
+				PrintKeys(&self.keys),
+				PrintChildKeys(&self.child_keys),
+			);
+		}
+	}
+}
+
+impl SubscriberSink {
+	fn new(subs_id: SubscriberId, keys: Keys, child_keys: ChildKeys) -> Self {
+		Self { subs_id, keys, child_keys, was_triggered: false }
+	}
+
+	fn render_notification<Hash>(
+		&mut self,
+		hash: Hash,
+		changes: Arc<[(StorageKey, Option<StorageData>)]>,
+		child_changes: Arc<[(StorageKey, Vec<(StorageKey, Option<StorageData>)>)]>,
+	) -> StorageNotification<Hash> {
+		self.was_triggered = true;
+
+		let storage_change_set = StorageChangeSet {
+			changes,
+			child_changes,
+			filter: self.keys.clone(),
+			child_filters: self.child_keys.clone(),
+		};
+
+		(hash, storage_change_set)
+	}
+}
+
+impl Registry {
+	pub(super) fn new(prometheus_registry: Option<PrometheusRegistry>) -> Self {
+		let metrics = prometheus_registry.and_then(|r| {
+			CounterVec::new(
+				Opts::new(
+					"substrate_storage_notification_subscribers",
+					"Number of subscribers in storage notification sytem",
+				),
+				&["action"], // added | removed
+			)
+			.and_then(|g| register(g, &r))
+			.ok()
+		});
+
+		Registry { metrics, ..Default::default() }
+	}
+}
+
+impl Default for Registry {
+	fn default() -> Self {
+		Self {
+			metrics: Default::default(),
+			wildcard_listeners: Default::default(),
+			listeners: Default::default(),
+			child_listeners: Default::default(),
+			sinks: Default::default(),
+		}
+	}
+}
+
+impl Unsubscribe for Registry {
+	fn unsubscribe(&mut self, subs_id: &SubscriberId) {
+		let _ = self.remove_subscriber(*subs_id);
+	}
+}
+
+impl<'a> Subscribe<SubscribeOp<'a>> for Registry {
+	fn subscribe(&mut self, subs_op: SubscribeOp<'a>, subs_id: SubscriberId) {
+		let SubscribeOp { filter_keys, filter_child_keys } = subs_op;
+
+		let keys = helpers::listen_from(
+			subs_id,
+			filter_keys.as_ref(),
+			&mut self.listeners,
+			&mut self.wildcard_listeners,
+		);
+
+		let child_keys = filter_child_keys.map(|filter_child_keys| {
+			filter_child_keys
+				.iter()
+				.map(|(c_key, o_keys)| {
+					let (c_listeners, c_wildcards) =
+						self.child_listeners.entry(c_key.clone()).or_default();
+
+					(
+						c_key.clone(),
+						helpers::listen_from(
+							subs_id,
+							o_keys.as_ref(),
+							&mut *c_listeners,
+							&mut *c_wildcards,
+						),
+					)
+				})
+				.collect()
+		});
+
+		if let Some(m) = self.metrics.as_ref() {
+			m.with_label_values(&[&"added"]).inc();
+		}
+
+		assert!(self.sinks.insert(subs_id, SubscriberSink::new(subs_id, keys, child_keys)).is_none(), "
+			Each `subs_id` is taken from `self.id_sequence.next_id()`.
+			If we have a duplicate key here, it's either the implementation of `IDSequence` was broken, or we've overflowed `u64`.
+			We are not likely to overflow an `u64`.");
```

Comment:
This proof is not correct, because you don't know how `subs_id` was generated.

Yes, I know it is done here in the pr, but that could be broken easily in some later refactoring.